File: ruby/secret-handshake/secret_handshake.rb
class SecretHandshake
  ANSWER_LIST = ['wink', 'double blink', 'close your eyes', 'jump']
  def initialize(input)
    @input = input
  end

  def commands
    answers = []
    answer_indexes = @input.to_s(2).chars.reverse.map.with_index { |e, i| e == '1' ? i : nil }.compact rescue nil
    return answers if answer_indexes.nil?
    answer_indexes = answer_indexes.reverse if answer_indexes[-1] == 4
    answer_indexes.each { |e| answers << ANSWER_LIST[e] }
    answers.compact
  end

end

File: python/secret-handshake/secret_handshake.py
COMMANDS = {
    'wink': 1,
    'double blink': 2,
    'close your eyes': 4,
    'jump': 8
}

def commands(number):
    commands = [k for k, v in COMMANDS.items() if v & number]
    return commands[::-1] if number & 16 else commands

def secret_code(actions):
    binary_list = [COMMANDS[act] for act in actions]
    return sum(binary_list) if binary_list[0] <= binary_list[-1] else 16 + sum(binary_list)

